src/main/java/org/folio/services/domainevent/AbstractDomainEventPublisher.java

```diff
@@ -0,0 +1,136 @@
+package org.folio.services.domainevent;
+
+import static io.vertx.core.Future.succeededFuture;
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+import static org.apache.logging.log4j.LogManager.getLogger;
+import static org.folio.rest.support.ResponseUtil.isCreateSuccessResponse;
+import static org.folio.rest.support.ResponseUtil.isDeleteSuccessResponse;
+import static org.folio.rest.support.ResponseUtil.isUpdateSuccessResponse;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.commons.lang3.tuple.Triple;
+import org.apache.logging.log4j.Logger;
+import org.folio.persist.AbstractRepository;
+import org.folio.services.batch.BatchOperationContext;
+
+import io.vertx.core.Future;
+
+abstract class AbstractDomainEventPublisher<DomainType, EventType> {
+  private static final Logger log = getLogger(AbstractDomainEventPublisher.class);
+
+  protected final AbstractRepository<DomainType> repository;
+  protected final CommonDomainEventPublisher<EventType> domainEventService;
+
+  public AbstractDomainEventPublisher(AbstractRepository<DomainType> repository,
+    CommonDomainEventPublisher<EventType> domainEventService) {
+
+    this.repository = repository;
+    this.domainEventService = domainEventService;
+  }
+
+  public Function<Response, Future<Response>> publishUpdated(DomainType oldRecord) {
+    return response -> {
+      if (!isUpdateSuccessResponse(response)) {
+        log.warn("Record update failed, skipping event publishing");
+        return succeededFuture(response);
+      }
```

Comment:
Just a `warn` in the log?

Reply:
At this point the response should forward error to the client and RMB should log all exceptions, the only think that we can is to log a warn that event sending is skipped and return the response so it will be sent to the client. 

What do you think?

Reply:
By looking at the response, does the client know that saving to DB was a success but domain event publishing failed? Also, when this happens, db and elastic search will not be in-sync anymore. How do we handle this?

Reply:
`if (!isUpdateSuccessResponse(response)) {` means that save to DB is unsuccessful. 

If this happens that DB operation is success but event publish is not then the error will be propagated to the client, and probably clients have to handle it manually. 

We can handle it with transaction, but it will be a significant complication of the current code.

What do you think?

CC: @MikhailFokanov

Reply:
Got it. If db fails to save, there will be no event sent. The log statement is sufficient in this case. I agree it is a hard decision whether we should consider db save and elastic search update in the same transaction scope. If not, how do we make those in sync. Thanks.

Reply:
>  I agree it is a hard decision whether we should consider db save and elastic search update in the same transaction scope.

They really can't be (even with the major pain of distributed transactions), there will likely always be a way for compensating actions could not happen.

Reply:
My take on it was that if someone (or some service) receive an error response due to failing of saving to the database, he (it) will manage to do appropriate actions (it is how it works (should work) right now).
If the database saving fails, the message won't be sent. So ES and DB are in sync.
If saving to database is succeed, and for some misty reason (not the likely case, as Kafka is cluster and fault-tolerant) sending of message fails, user should get the same error. And manage to retry his request (as it is for the first case). So ES and DB should be also in sync. AFAIK, for the inventory the writing of the same data to the database don't break anything (during the retry).

Reply:
The issue is that the caller can put system out of sync. It should be the microservice's responsibility to maintain its own consistent state. The retry (if caller is responsible) might work for update, but not sure about create/delete.

Reply:
> If saving to database is succeed, and for some misty reason (not the likely case, as Kafka is cluster and fault-tolerant) sending of message fails, user should get the same error. And manage to retry his request (as it is for the first case).

This suggests to me that we are relying on the client to retry in order for the system to become consistent. 

Am I interpreting this correctly?

Reply:
Hi guys, I think it is a general FOLIO challenge rather then an issue for this particular case. There are at least few more APIs that can easily break DB in inconsistent state: for example circulation check-in/out, requests, data import (I imagine) etc. 

So we probably need a community agreement for such a problem. If it you still think that it is critical problem, than we should investigate if we are able to use DB transactions for this (actually we can, but RMB has some limited toolset for this: for example there is no getById method with connection - it is needed to read records within the same transaction before commit. And this limited toolset requires introducing some custom methods in mod-inventory - which should be very well tested).

What do you think?

Reply:
https://issues.folio.org/browse/MODINVSTOR-663